src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClientManager.cs

```diff
@@ -57,19 +111,106 @@ internal ConfigurationClientManager(IList<ConfigurationClientWrapper> clients)
             _clients = clients;
         }
 
-        public IEnumerable<ConfigurationClient> GetAvailableClients(DateTimeOffset time)
+        public async IAsyncEnumerable<ConfigurationClient> GetAvailableClients([EnumeratorCancellation] CancellationToken cancellationToken)
         {
-            return _clients.Where(client => client.BackoffEndTime <= time).Select(c => c.Client).ToList();
-        }
+            DateTimeOffset now = DateTimeOffset.UtcNow;
+            DateTimeOffset minimalRefreshTime = _lastFallbackClientRefreshAttempt + MinimalClientRefreshInterval;
+            Task task = null;
+
+            // Principle of refreshing fallback clients:
+            // 1.Perform initial refresh attempt to query available clients on app start-up.
+            // 2.Refreshing when cached fallback clients have expired, FallbackClientRefreshExpireInterval is due.
+            // 3.At least wait MinimalClientRefreshInterval between two attempt to ensure not perform refreshing too often.
+
+            if (_replicaDiscoveryEnabled &&
+                now >= minimalRefreshTime && 
+                (_dynamicClients == null ||
+                now >= _lastFallbackClientRefresh + FallbackClientRefreshExpireInterval))
+            {
+                _lastFallbackClientRefreshAttempt = now;
+
+                var cts = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
+                cts.CancelAfter(SrvLookupTimeout);
+
+                task = RefreshFallbackClients(cts.Token);
+
+                // Observe cancellation if it occurs
+                _ = task.ObserveCancellation(_logger);
+
+                _ = task.ContinueWith(t =>
+                {
+                    cts.Dispose();
+                });
+            }
 
+            foreach (ConfigurationClientWrapper clientWrapper in _clients)
+            {
+                if (clientWrapper.BackoffEndTime <= now)
+                {
+                    yield return clientWrapper.Client;
+                }
+            }
+
+            // No need to continue if replica discovery is not enabled.
+            if (!_replicaDiscoveryEnabled)
+            {
+                yield break;
+            }
+
+            if (_dynamicClients != null)
+            {
+                foreach (ConfigurationClientWrapper clientWrapper in _dynamicClients)
+                {
+                    if (clientWrapper.BackoffEndTime <= now)
+                    {
+                        yield return clientWrapper.Client;
+                    }
+                }
+            }
+
+            // All static and dynamic clients exhausted, check if previously unknown
+            // dynamic client available
+            if (now >= minimalRefreshTime)
+            {
+                if (task != null)
+                {
+                    await task.ConfigureAwait(false);

```

Comment:
That's a problem because this can cancel due to our own internal cancellation, even though the caller didn't cancel.

Reply:
```suggestion
                    try
                    {
                        await task.ConfigureAwait(false);
                    }
                    catch (OperationCanceledException) when (!cancellationToken.IsCancellationRequested)
                    {
                        await RefreshFallbackClients(cancellationToken).ConfigureAwait(false);
                    }
```